services/storage/src/simcore_service_storage/simcore_s3_dsm.py

```diff
@@ -383,16 +387,47 @@ async def complete_file_upload(
     async def create_file_download_link(
         self, user_id: UserID, file_id: StorageFileID, link_type: LinkType
     ) -> AnyUrl:
-        async with self.engine.acquire() as conn:
+        """
+        Cases:
+        1. the file_id maps 1:1 to `file_meta_data` (e.g. it is not a file inside a directory)
+        2. the file_id represents a file inside a directory (its root path maps 1:1 to a `file_meta_data` defined as a directory)
+
+        3. Raises FileNotFoundError if the file does not exist
+        4. Raises FileAccessRightError if the user does not have access to the file
+        """
+
+        async def _ensure_access_rights(
+            conn: SAConnection, storage_file_id: StorageFileID
+        ) -> None:
             can: AccessRights | None = await get_file_access_rights(
-                conn, user_id, file_id
+                conn, user_id, storage_file_id
             )
             if not can.read:
                 # NOTE: this is tricky. A user with read access can download and data!
                 # If write permission would be required, then shared projects as views cannot
                 # recover data in nodes (e.g. jupyter cannot pull work data)
                 #
-                raise FileAccessRightError(access_right="read", file_id=file_id)
+                raise FileAccessRightError(access_right="read", file_id=storage_file_id)
+
+        async def _get_link(s3_file_id: SimcoreS3FileID) -> AnyUrl:
```

Comment:
Thought: I am still not very fond of having all these function in there vs raw code. this is for sure less efficient, why not move these at least at the top of the file?